File: v1/src/main/java/com/api/v1/utils/customer/CustomerFinderUtil.java
package com.api.v1.utils.customer;

import com.api.v1.annotations.customer.SSN;
import com.api.v1.domain.customer.Customer;
import com.api.v1.domain.customer.CustomerRepository;
import com.api.v1.exceptions.customer.CustomerNotFoundException;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
public class CustomerFinderUtil {

    @Autowired
    private CustomerRepository repository;

    public Mono<Customer> find(@SSN String ssn) {
        return repository
                .findAll()
                .filter(e -> e.getSsn().equals(ssn))
                .singleOrEmpty()
                .switchIfEmpty(Mono.error(new CustomerNotFoundException(ssn)));
    }

}

File: kotlin/v2/src/main/kotlin/com/api/v2/customer/utils/CustomerFinderUtil.kt
package com.api.v2.customer.utils

import com.api.v2.customer.domain.Customer
import com.api.v2.customer.domain.CustomerRepository
import com.api.v2.customer.exceptions.CustomerNotFoundException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.withContext
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.stereotype.Component

@Component
class CustomerFinderUtil {

    @Autowired
    private lateinit var customerRepository: CustomerRepository

    suspend fun find(ssn: String): Customer {
        return withContext(Dispatchers.IO) {
            customerRepository
                .findAll()
                .filter { e -> e.ssn ==  ssn }
                .firstOrNull() ?: throw CustomerNotFoundException()
        }
    }

}